internal/controller/apicerts/apiservice_updater.go

```diff
@@ -60,7 +60,9 @@ func (c *apiServiceUpdaterController) Sync(ctx controllerlib.Context) error {
 	if notFound {
 		// The secret does not exist yet, so nothing to do.
 		klog.Info("apiServiceUpdaterController Sync found that the secret does not exist yet or was deleted")
-		return nil
+		//nolint: goerr113
+		return fmt.Errorf("apiServiceUpdaterController missing pre-requirements, secret %s/%s does not exist: %w",
```

Comment:
Hi @enj! Thanks for the PR! I had a question about your recommendations for how to return from controllers made in this PR.

Take this line for example. In this case the controller has decided that there is no work to do until the Secret is created (or recreated) by another party. So the controller does not need to be scheduled for sync again until the next time a Secret changes/creates/deletes or the next full resync, whichever happens first. Isn't this exactly what `return nil` means?

Reply:
This is probably a discussion better had via a higher bandwidth medium 😄 

In brief, the way I think about the `Syncer` interface contract is that per invocation:

>>> For the current key, have I successfully completed the task?
>>> If yes, return `nil`, else return some error to describe the problem.

For the `apiServiceUpdaterController`, the contact is (extra simplified because the queue has a singleton key):

>>> Have I updated the API service with the CA bundle required to make API aggregation work correctly?
>>> If yes, return `nil`, else return some error to describe the problem.

---

> In this case the controller has decided that there is no work to do until the Secret is created (or recreated) by another party.

I would instead describe this as:

>>> I cannot update the API service because I cannot locate the CA bundle, thus I must return an error.

If you ignore the log statement (which is not necessary if you return an error), one should communicate the inability to make forward progress.  `WithMaxRetries` can be used if you would prefer to not keep retrying infidelity, though in the case of a singleton queue I think it is more correct to just defer to the rate limiter.

> So the controller does not need to be scheduled for sync again until the next time a Secret changes/creates/deletes or the next full resync, whichever happens first.

In terms of the `Syncer`, I would avoid thinking of scheduling unless you are directly interacting with the `Queue`.

> Isn't this exactly what `return nil` means?

If `apiServiceUpdaterController` returns `nil`, I would expect the API service to be in a state that the API server could use it to successfully create a TLS connection to our service with no certificate errors.